mypy/test/data/check-optional.test

```diff
@@ -0,0 +1,163 @@
+-- Tests for strict Optional behavior
+
+[case testImplicitNoneType]
+x = None
+x()  # E: None not callable
+
+[case testExplicitNoneType]
+x = None  # type: None
+x()  # E: None not callable
+
+[case testNoneMemberOfOptional]
+from typing import Optional
+x = None  # type: Optional[int]
+
+[case testTypeMemberOfOptional]
+from typing import Optional
+x = 0  # type: Optional[int]
+
+[case testNoneNotMemberOfType]
+x = None  # type: int
+[out]
+main:1: error: Incompatible types in assignment (expression has type None, variable has type "int")
+
+[case testTypeNotMemberOfNone]
+x = 0  # type: None
+[out]
+main:1: error: Incompatible types in assignment (expression has type "int", variable has type None)
+
+[case testOptionalNotMemberOfType]
+from typing import Optional
+def f(a: int) -> None: pass
+x = None  # type:  Optional[int]
+f(x)  # E: Argument 1 to "f" has incompatible type "Optional[int]"; expected "int"
+
+[case testIsinstanceCases]
+from typing import Optional
+def f(a: int) -> None: pass
+def g(a: None) -> None: pass
+x = None  # type:  Optional[int]
+if isinstance(x, int):
+  f(x)
+else:
+  g(x)
+[builtins fixtures/isinstance.py]
+
+[case testIfCases]
+from typing import Optional
+def f(a: int) -> None: pass
+def g(a: None) -> None: pass
+x = None  # type:  Optional[int]
+if x:
+  f(x)
+else:
+  g(x)  # E: Argument 1 to "g" has incompatible type "Optional[int]"; expected None
+[builtins fixtures/bool.py]
+
+[case testIfNotCases]
+from typing import Optional
+def f(a: int) -> None: pass
+def g(a: None) -> None: pass
+x = None  # type:  Optional[int]
+if not x:
+  g(x)  # E: Argument 1 to "g" has incompatible type "Optional[int]"; expected None
+else:
+  f(x)
+  x()  # E: "int" not callable
+[builtins fixtures/bool.py]
+
+[case testIsNotNoneCases]
+from typing import Optional
+def f(a: int) -> None: pass
+def g(a: None) -> None: pass
+x = None  # type:  Optional[int]
+if x is not None:
+  f(x)
+  x()  # E: "int" not callable
+else:
+  g(x)
+  x()  # E: None not callable
+[builtins fixtures/bool.py]
+
+[case testIsNoneCases]
+from typing import Optional
+def f(a: int) -> None: pass
+def g(a: None) -> None: pass
+x = None  # type:  Optional[int]
+if x is None:
+  g(x)
+else:
+  f(x)
+[builtins fixtures/bool.py]
+
+[case testLambdaReturningNone]
+f = lambda: None
+x = f()
+
+[case testFunctionStillVoid]
+def f() -> None: pass
+f()
+x = f()  # E: "f" does not return a value
+
+[case testNoneArgumentType]
+def f(x: None) -> None: pass
+f(None)
+
+[case testInferOptionalFromDefaultNone]
+def f(x: int = None) -> None:
+  x + 1  # E: Unsupported left operand type for + (some union)
```

Comment:
It certainly would be nice if `Optional[x]` would never be shown as "some union" :-/